src/internal/logger.hh

```diff
@@ -0,0 +1,38 @@
+#include "zarr.h"
+
+#include <string>
+
+class Logger
+{
+  public:
+    static void set_log_level(LogLevel level);
+    static LogLevel get_log_level();
+
+    static std::string log(LogLevel level,
+                           const char* file,
+                           int line,
+                           const char* func,
+                           const char* format,
+                           ...);
+
+  private:
+    static LogLevel current_level;
+};
+
+#define LOG_DEBUG(...)                                                         \
+    Logger::log(LogLevel_Debug, __FILE__, __LINE__, __func__, __VA_ARGS__)
+#define LOG_INFO(...)                                                          \
+    Logger::log(LogLevel_Info, __FILE__, __LINE__, __func__, __VA_ARGS__)
+#define LOG_WARNING(...)                                                       \
+    Logger::log(LogLevel_Warning, __FILE__, __LINE__, __func__, __VA_ARGS__)
+#define LOG_ERROR(...)                                                         \
+    Logger::log(LogLevel_Error, __FILE__, __LINE__, __func__, __VA_ARGS__)
+
+#define EXPECT(e, ...)                                                         \
+    do {                                                                       \
+        if (!(e)) {                                                            \
+            const std::string __err = LOG_ERROR(__VA_ARGS__);                  \
+            throw std::runtime_error(__err);                                   \
+        }                                                                      \
+    } while (0)
+#define CHECK(e) EXPECT(e, "Expression evaluated as false:\n\t%s", #e)
```

Comment:
I feel like all these macros can be replaced with static member functions, for example:
```cpp
template <typename... Args>
static void debug(const char* format, Args... args) {
   log(LogLevel_Debug, __FILE__, __LINE__, __func__, format, args...);
}
```
This approach improves type checking (which you get from variadic templates), and it avoids all the issues that come with macro expansions.